### PR TITLE
HPCC-21958 Implement value sets containing substring filters

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman initial PR - for smoketest.
It should be ready for merging, but it is much larger than would normally be included in a point release, so needs very careful review and debate about which branch to target.

File: ecl/hql/hqlfilter.cpp [hunk 1]
```diff
@@ -1902,40 +1904,42 @@ IValueSet * FilterExtractor::createValueSetCompareExpr(IHqlExpression * selector
     if (!createConstantField(compareValue, selector, normalized))
         throwTooComplex(expr);
 
+    //TBD: Support substring matches
```
Comment (richardkchapman) [inline ecl/hql/hqlfilter.cpp]: Is this comment still valid?

File: ecl/hql/hqlfilter.cpp [hunk 2]
```diff
@@ -1902,40 +1904,42 @@ IValueSet * FilterExtractor::createValueSetCompareExpr(IHqlExpression * selector
...
     case no_ne:
         values->addAll();
         if (!truncated)
-            values->killRawRange(compareRaw, compareRaw);
+            values->killRawRangeEx(compareRaw, compareRaw, subLength);
         break;
     case no_le:
     {
-        Owned<IValueTransition> upper = values->createRawTransition(CMPle, compareRaw);
+        Owned<IValueTransition> upper = values->createRawTransitionEx(CMPle, compareRaw, subLength); //MORE Ex or call addRawRange instead.
```
Comment (richardkchapman) [inline ecl/hql/hqlfilter.cpp]: Is this comment still valid?

File: ecl/hql/hqlfilter.cpp [hunk 3]
```diff
@@ -1902,40 +1904,42 @@ IValueSet * FilterExtractor::createValueSetCompareExpr(IHqlExpression * selector
...
         values->addRange(nullptr, upper);
         break;
     }
     case no_lt:
     {
-        Owned<IValueTransition> upper = values->createRawTransition(truncated ? CMPle: CMPlt, compareRaw);
+        Owned<IValueTransition> upper = values->createRawTransitionEx(truncated ? CMPle: CMPlt, compareRaw, subLength); // MORE: Ex
         values->addRange(nullptr, upper);
         break;
```
Comment (richardkchapman) [inline ecl/hql/hqlfilter.cpp]: Is this comment still valid?

File: ecl/hqlcpp/hqlcfilter.cpp [hunk 4]
```diff
@@ -328,6 +342,36 @@ void CppFilterExtractor::createStringSet(BuildCtx & ctx, const char * target, un
...
+    case no_rangeto:
+        limit = range->queryChild(0);
+        break;
+    case no_range:
+        assertex(matchesConstValue(range->queryChild(0), 1));
+        limit = range->queryChild(1);
+        break;
+    case no_constant:
+        limit = range;
+        assertex(matchesConstValue(range, 1));
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlcfilter.cpp]: I don't understand this line

Comment (richardkchapman) [reply ecl/hqlcpp/hqlcfilter.cpp]: I think it means that we should only be here with leading substrings, including leading single-char substrings. Which makes sense (and this code appears to have moved from elsewhere)

File: ecl/hql/hqlfilter.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlfilter.cpp]: Yes - I should open a jira.  It will be needed once this code is used for filtering in esp.

Comment (ghalliday) [reply ecl/hql/hqlfilter.cpp]: HPCC-22017

File: rtl/eclrtl/rtlfield.cpp [hunk 5]
```diff
@@ -291,6 +320,72 @@ void RtlTypeInfoBase::getUtf8ViaString(size32_t & resultLen, char * & result, co
...
+                //but currently the index code uses memcmp() so they need to be initialised
+                unsigned copyLen = std::min(length, subLength);
+                copySize = copyLen;
+                memcpy(dst, value, copySize);
+                char padChar;
+                memset(dst+copySize, fill, minSize-copySize);
+            }
+            else
+            {
+                assertex(!inclusive);
```
Comment (ghalliday) [inline rtl/eclrtl/rtlfield.cpp]: test is inverted

File: rtl/eclrtl/rtlfield.cpp [hunk 6]
```diff
@@ -291,6 +320,72 @@ void RtlTypeInfoBase::getUtf8ViaString(size32_t & resultLen, char * & result, co
...
+        if (subLength != MatchFullString)
+        {
+            if (isFixedSize())
+            {
+                //For a substring comparison, clear the bytes after subLength - eventually they should never be compared
+                //but currently the index code uses memcmp() so they need to be initialised
+                unsigned copyLen = std::min(length, subLength);
+                copySize = copyLen;
+                memcpy(dst, value, copySize);
+                char padChar;
```
Comment (ghalliday) [inline rtl/eclrtl/rtlfield.cpp]: unused

File: rtl/eclrtl/rtlfield.cpp [hunk 7]
```diff
@@ -2325,6 +2573,66 @@ int RtlUnicodeTypeInfo::compare(const byte * left, const byte * right) const
...
+    byte *dst = (byte *) buffer;
+    size32_t minSize = getMinSize();
+    if (value)
+    {
+        assertex(inclusive);
+        if (subLength != MatchFullString)
+        {
+            if (isFixedSize())
+            {
+                //Only support a lower bound for unicode substrings - filled with spavces
```
Comment (ghalliday) [inline rtl/eclrtl/rtlfield.cpp]: typo spaces

File: rtl/eclrtl/rtlnktest.cpp [hunk 8]
```diff
@@ -186,6 +186,8 @@ class RawRowCompare : public ICompare
 class ValueSetTest : public CppUnit::TestFixture
 {
     CPPUNIT_TEST_SUITE(ValueSetTest);
+        CPPUNIT_TEST(testBounds);
+/*        CPPUNIT_TEST(testSubStrings);
```
Comment (ghalliday) [inline rtl/eclrtl/rtlnktest.cpp]: commented out by mistake

File: testing/ecl/setup/setup_fetch.ecl [hunk 9]
```diff
@@ -77,6 +77,9 @@ output(sort(twoways,record,local),,DG_FetchFilePreloadIndexedName,OVERWRITE);
 
 #END
 
+output('Hello');
```
Comment (ghalliday) [inline testing/ecl/setup/setup_fetch.ecl]: old file modified by mistake

**Review by ghalliday — COMMENTED** [review]
A few minor comments

**Review by shamser — APPROVED** [review]
Code reviewed.  Looks good.

**Pull request merged**